cmd/mv-main.go

```diff
@@ -0,0 +1,356 @@
+
+  13. Move a text file to an object storage and assign REDUCED_REDUNDANCY storage-class to the uploaded object.
+      {{.Prompt}} {{.HelpName}} --storage-class REDUCED_REDUNDANCY myobject.txt play/mybucket
+
+  14. Move a text file to an object storage and create or resume copy session.
+      {{.Prompt}} {{.HelpName}} --recursive --continue dir/ play/mybucket
+
+  15. Move a text file to an object storage and preserve the file system attribute as metadata.
+      {{.Prompt}} {{.HelpName}} -a myobject.txt play/mybucket
+
+  16. Move a text file to an object storage with object lock mode set to 'GOVERNANCE' with retention date.
+      {{.Prompt}} {{.HelpName}} --attr "x-amz-object-lock-mode=GOVERNANCE;x-amz-object-lock-retain-until-date=2020-01-11T01:57:02Z" locked.txt play/locked-bucket/
+`,
+}
+
+type removeClientInfo struct {
+	client    Client
+	contentCh chan *ClientContent
+	errorCh   <-chan *probe.Error
+}
+
+type removeManager struct {
+	removeMap      map[string]*removeClientInfo
+	removeMapMutex sync.RWMutex
+	doneCh         chan struct{}
+	isClosed       int32
+	wg             sync.WaitGroup
+}
+
+func (rm *removeManager) readErrors(errorCh <-chan *probe.Error, targetURL string) {
+	rm.wg.Add(1)
+	go func() {
+		defer rm.wg.Done()
+		var stop bool
+		for !stop {
+			select {
+			case pErr, ok := <-errorCh:
+				if ok {
+					errorIf(pErr.Trace(targetURL), "Failed to remove in`"+targetURL+"`.")
+				}
+			case <-rm.doneCh:
+				stop = true
+			}
+		}
+
+		for pErr := range errorCh {
+			if pErr != nil {
+				errorIf(pErr.Trace(targetURL), "Failed to remove in `"+targetURL+"`.")
+			}
+		}
+	}()
+}
+
+func (rm *removeManager) add(targetAlias, targetURL string) {
+	if atomic.LoadInt32(&rm.isClosed) != 0 {
+		return
+	}
+
+	rm.removeMapMutex.RLock()
+	clientInfo := rm.removeMap[targetAlias]
+	rm.removeMapMutex.RUnlock()
+
+	if clientInfo == nil {
+		client, pErr := newClientFromAlias(targetAlias, targetURL)
+		if pErr != nil {
+			errorIf(pErr.Trace(targetURL), "Invalid argument `"+targetURL+"`.")
+			return
+		}
+
+		contentCh := make(chan *ClientContent)
+		errorCh := client.Remove(false, false, contentCh)
+		rm.readErrors(errorCh, targetURL)
+
+		clientInfo = &removeClientInfo{
+			client:    client,
+			contentCh: contentCh,
+			errorCh:   errorCh,
+		}
+
+		rm.removeMapMutex.Lock()
+		rm.removeMap[targetAlias] = clientInfo
+		rm.removeMapMutex.Unlock()
+	}
+
+	go func() {
+		clientInfo.contentCh <- &ClientContent{URL: *newClientURL(targetURL)}
+	}()
+}
+
+func (rm *removeManager) close() {
+	atomic.StoreInt32(&rm.isClosed, 1)
+	rm.removeMapMutex.Lock()
+	defer rm.removeMapMutex.Unlock()
+
+	for _, clientInfo := range rm.removeMap {
+		close(clientInfo.contentCh)
+	}
+
+	close(rm.doneCh)
+
+	rm.wg.Wait()
+}
+
+var rmManager = &removeManager{
+	removeMap: make(map[string]*removeClientInfo),
+	doneCh:    make(chan struct{}),
+}
+
+func bgRemove(url string) {
```

Comment:
It would be useful to have comments explaining why we use a different approach when `targetAlias` is empty .

Reply:
done